rust/connlib/tunnel/src/gateway/nat_table.rs

```
@@ -16,32 +16,66 @@ use std::time::{Duration, Instant};
 /// Thus, purely an L3 NAT would not be sufficient as it would be impossible to map back to the proxy IP.
 #[derive(Default, Debug)]
 pub(crate) struct NatTable {
-    pub(crate) table: BiMap<(Protocol, IpAddr), (Protocol, IpAddr)>,
-    pub(crate) last_seen: BTreeMap<(Protocol, IpAddr), Instant>,
+    table: BiMap<Inside, Outside>,
+    state_by_inside: BTreeMap<Inside, EntryState>,
 
     // We don't bother with proactively freeing this because a single entry is only ~20 bytes and it gets cleanup once the connection to the client goes away.
-    expired: HashSet<(Protocol, IpAddr)>,
+    expired: HashSet<Outside>,
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Clone, Copy)]
+struct Inside(Protocol, IpAddr);
+
+impl Inside {
+    fn into_inner(self) -> (Protocol, IpAddr) {
+        (self.0, self.1)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Clone, Copy)]
+struct Outside(Protocol, IpAddr);
+
+impl Outside {
+    fn into_inner(self) -> (Protocol, IpAddr) {
+        (self.0, self.1)
+    }
 }
 
 pub(crate) const TCP_TTL: Duration = Duration::from_secs(60 * 60 * 2);
 pub(crate) const UDP_TTL: Duration = Duration::from_secs(60 * 2);
 pub(crate) const ICMP_TTL: Duration = Duration::from_secs(60 * 2);
 
+pub(crate) const UNCONFIRMED_TTL: Duration = Duration::from_secs(60);
+
 impl NatTable {
     pub(crate) fn handle_timeout(&mut self, now: Instant) {
-        for (outside, e) in self.last_seen.iter() {
-            let ttl = match outside.0 {
-                Protocol::Tcp(_) => TCP_TTL,
-                Protocol::Udp(_) => UDP_TTL,
-                Protocol::Icmp(_) => ICMP_TTL,
+        let expired = self.state_by_inside.extract_if(.., |inside, state| {
+            state
+                .remove_at(inside.0)
+                .is_some_and(|remove_at| now >= remove_at)
+        });
+
+        for (inside, state) in expired {
+            let Some((_, outside)) = self.table.remove_by_left(&inside) else {
+                continue;
```

can these 2 data structures - `table ` and `state_by_inside` get out of sync?
Slightly concerned about lack of atomicity when removing entries and silently ignoring if there wasn't a matching entry in the `table`

---

They should never get out of sync but it isn't totally protected against. We'd have to extract a dedicated data structure for that which allows storing an additional T with a `BiMap`.

---

Not having an entry in `table` is fine. What would be more problematic is having an entry in `table` without also having one in `state_by_insinde`.